tests/unit/tf/models/test_retrieval.py

```diff
@@ -997,3 +997,51 @@ def test_youtube_dnn_topk_evaluation(sequence_testing_data: Dataset, run_eagerly
 
     metrics = topk_model.evaluate(dataloader, return_dict=True, pre=predict_next)
     assert all([metric >= 0 for metric in metrics.values()])
+
+
+@pytest.mark.parametrize(
+    "pos_class_weight", [1, tf.random.uniform(shape=(1000,)), "user_item_intentions"]
+)
+@pytest.mark.parametrize("neg_class_weight", [0.1, tf.random.uniform(shape=(1000,))])
+def test_two_tower_v2_with_contrastive_sample_weight(
+    ecommerce_data: Dataset, pos_class_weight, neg_class_weight
+):
+    from merlin.models.tf.experimental.sample_weight import ContrastiveSampleWeight
+
+    tower_dim = 64
+    data = ecommerce_data
+    input_features = ["user_categories", "item_id"]
+    if isinstance(pos_class_weight, str):
+        input_features += [pos_class_weight]
+    data.schema = data.schema.select_by_name(input_features)
+
+    user_schema = data.schema.select_by_tag(Tags.USER)
+    user_inputs = mm.InputBlockV2(user_schema)
+    query = mm.Encoder(user_inputs, mm.MLPBlock([128, tower_dim]))
+
+    item_schema = data.schema.select_by_tag(Tags.ITEM)
+    item_inputs = mm.InputBlockV2(item_schema)
+    candidate = mm.Encoder(item_inputs, mm.MLPBlock([128, tower_dim]))
+
+    output = mm.ContrastiveOutput(
+        DotProduct(),
+        post=None,
+        schema=data.schema.select_by_tag(Tags.ITEM_ID),
+        negative_samplers="in-batch",
+        store_negative_ids=True,
+    )
+    model = mm.TwoTowerModelV2(query, candidate, outputs=output, schema=data.schema)
+    model.compile(run_eagerly=False, loss="binary_crossentropy")
+
+    batch = mm.sample_batch(data, batch_size=16)
+    # loss without sample weight
+    metrics1 = model.train_step(batch)
```

Comment:
I think calling `train_step()` twice with the same batch results in different `loss` because it does parameters optimization under-the-hood. You could change it to `test_step()` if you add support to evaluation mode in `ContrastiveSampleWeight`.

Reply:
You’re right, I fixed it by using `test_step` instead.